crates/sui-types/src/message_envelope.rs

```diff
@@ -106,7 +106,8 @@ pub trait Message {
 
     /// Verify that the message is from the correct epoch (e.g. for CertifiedCheckpointSummary
     /// we verify that the checkpoint is from the same epoch as the committee signatures).
-    fn verify_epoch(&self, epoch: EpochId) -> SuiResult;
+    fn verify_epoch(&self, epoch: EpochId, upper_bound_for_max_epoch: Option<EpochId>)
```

Comment:
Is `upper_bound_for_max_epoch` only used for Transaction? If so I am not convinced it belongs to the general message trait.

Reply:
theres actually no concrete type at a few places where the verify_epoch is evoked, i can only access it via the Message trait: 
```
impl<T> Envelope<T, AuthoritySignInfo>
where
    T: Message + AuthenticatedMessage + Serialize,
{
    pub fn verify_signatures_authenticated(
        &self,
        committee: &Committee,
        verify_params: &VerifyParams,
    ) -> SuiResult {
        self.data.verify_epoch(self.auth_sig().epoch, None)?;
```

and 
```
impl<T> Envelope<T, AuthoritySignInfo>
where
    T: Message + UnauthenticatedMessage + Serialize,
{
    pub fn verify_authority_signatures(&self, committee: &Committee) -> SuiResult {
        self.data.verify_epoch(self.auth_sig().epoch, None)?;
```

Reply:
there are 4 concrete types for the message trait, 2 out of them actually just return Ok(()) noop, im guessing its a bit messy to remove the trait?

Reply:
I think that just means `verify_epoch` shouldn't be a trait API for `Message`. SenderSignedData and SignedCheckpointSummary should have its own function to check this.

Reply:
`verify_max_epoch_for_all_sigs` i added this call for SenderSignedData and this is only triggered inside verify_tx. the other other implementations like Checkpoint, TransactionEffects etc should not change its interface

```
                signed_tx.verify_epoch(self.committee.epoch())?;
                signed_tx.verify_max_epoch_for_all_sigs(
                    self.committee.epoch(),
                    self.zk_login_params.zklogin_max_epoch_upper_bound,
                )?;
```

however it 2 iterations of the for loop instead of 1. lmk if there is better idea how to refactor this @lxfind  @mystenmark